curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -216,55 +225,72 @@ S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
 
 int S3CompactWorkQueueImpl::ReadFullChunk(
     const std::list<struct S3CompactWorkQueueImpl::Node>& validList,
-    uint64_t fsId, uint64_t inodeId, uint64_t blockSize, std::string* fullChunk,
-    uint64_t* newChunkid, uint64_t* newCompaction, S3Adapter* s3adapter) {
+    uint64_t fsId, uint64_t inodeId, uint64_t blockSize, uint64_t chunkSize,
+    std::string* fullChunk, uint64_t* newChunkid, uint64_t* newCompaction,
+    S3Adapter* s3adapter) {
     for (auto curr = validList.begin(); curr != validList.end();) {
         auto next = std::next(curr);
         if (curr->zero) {
-            fullChunk->append(curr->end - curr->begin, '\0');
+            fullChunk->append(curr->end - curr->begin + 1, '\0');
             curr = next;
             continue;
         }
-        // block index is relatively in a s3chunkinfo
-        for (uint64_t index = (curr->begin - curr->chunkoff) / blockSize;
-             index * blockSize + curr->chunkoff < curr->end; index++) {
+
+        VLOG(9) << "s3compact: read [" << curr->begin << "-" << curr->end
+                << "]";
+        uint64_t startIndex =
```

Comment:
curr->begin - curr->begin / chunkSize * chunkSize is equivalent to curr->begin - curr->begin / chunkSize * chunkSize
startIndex = begin % chunksize / blocksize?

Reply:
add posRoundDown = pos/chunkSize*chunkSize to simplify code. it is reused many times.
%  = pos - posRoundDown